### PR TITLE
Quick Fixes

### LINKED ISSUE #16: getproperties() takes excessive amount of time
When running the getproperties function. The call takes an excessive amount of time. This isn't a problem when just using the REPL and exploring, but any type of data acquisition script will likely suffer catastrophic timing results due to the use of getproperties in every read and write call.

Recommend either a struct for a task (allows people to break things by using native calls on a task instead of high level calls) or just multiple dispatch to have a more focused ccall specifically for reads and writes.

```
julia> @time getproperties()
  2.729390 seconds (859.03 k allocations: 53.213 MiB, 0.43% gc time)
Dict{String,Tuple{Any,Bool}} with 7 entries:
  "Tasks"              => (SubString{String}["CS", "MOSI", "MyAccelerationTask", "MyVoltageTask", "SCK", "SPI", "testVoltageTask"], false)
  "NIDAQUpdateVersion" => (0x00000000, false)
  "NIDAQMinorVersion"  => (0x00000006, false)
  "Scales"             => (SubString{String}[""], false)
  "DevNames"           => (SubString{String}["Dev1"], false)
  "GlobalChans"        => (SubString{String}[""], false)
  "NIDAQMajorVersion"  => (0x00000012, false)

julia> @time t = analog_input("Dev1/ai0")
  0.378259 seconds (509.50 k allocations: 25.669 MiB, 1.47% gc time)
NIDAQ.AITask(Ptr{Nothing} @0x0000000030c0f900)

julia> @time start(t)
  0.007996 seconds (2.82 k allocations: 158.683 KiB)

julia> @time read(t, Float64, 1)
  2.931029 seconds (1.33 M allocations: 76.679 MiB, 0.49% gc time)
1-element Array{Float64,1}:
 -0.0006444593315980392

julia> using Profile

julia> @profile read(t, Float64, 1)
1-element Array{Float64,1}:
 1.5828057680061097e-8

julia> Profile.print()
1    .\client.jl:436; _start()
 1 .\client.jl:284; exec_options(::Base.JLOptions)
  1 .\client.jl:346; run_main_repl(::Bool, ::Bool, ::Bool, ::Bool, ::Bool)
   1 .\essentials.jl:741; invokelatest
    1 .\essentials.jl:742; #invokelatest#1
     1 .\client.jl:362; (::getfield(Base, Symbol("##734#736")){Bool,Bool,Bool,Bool})(::Module)
      1 ...bot\worker\package_win64\build\usr\share\julia\stdlib\v1.1\REPL\src\REPL.jl:192; run_repl(::REPL.AbstractREPL, ::Any)
       1 ...bot\worker\package_win64\build\usr\share\julia\stdlib\v1.1\REPL\src\REPL.jl:1035; run_frontend(::REPL.LineEditREPL, ::REPL.REPLBackendRef)
        1 ...orker\package_win64\build\usr\share\julia\stdlib\v1.1\REPL\src\LineEdit.jl:2273; run_interface(::REPL.Terminals.TextTerminal, ::REPL.LineEdit.ModalInterface, ::REPL.LineEdit.MIState)
         1 .\essentials.jl:741; invokelatest
          1 .\essentials.jl:742; #invokelatest#1
           1 ...t\worker\package_win64\build\usr\share\julia\stdlib\v1.1\REPL\src\REPL.jl:714; (::getfield(REPL, Symbol("#do_respond#38")){Bool,getfield(REPL, Symbol("##48#57")){REPL.LineEditREPL,REPL.REPLHi...
            1 ...t\worker\package_win64\build\usr\share\julia\stdlib\v1.1\REPL\src\REPL.jl:140; print_response(::REPL.AbstractREPL, ::Any, ::Any, ::Bool, ::Bool)
             1 ...t\worker\package_win64\build\usr\share\julia\stdlib\v1.1\REPL\src\REPL.jl:155; print_response(::IO, ::Any, ::Any, ::Bool, ::Bool, ::Any)
              1 .\essentials.jl:741; invokelatest
               1 .\essentials.jl:742; #invokelatest#1
                1 .\multimedia.jl:287; display(::Any)
                 1 ...worker\package_win64\build\usr\share\julia\stdlib\v1.1\REPL\src\REPL.jl:135; display(::REPL.REPLDisplay, ::Any)
                  1 ...worker\package_win64\build\usr\share\julia\stdlib\v1.1\REPL\src\REPL.jl:131; display(::REPL.REPLDisplay, ::MIME{Symbol("text/plain")}, ::Any)
                   1 .\arrayshow.jl:316; show(::IOContext{REPL.Terminals.TTYTerminal}, ::MIME{Symbol("text/plain")}, ::Array{Float64,1})
                    1 .\show.jl:1837; summary
                     1 .\show.jl:1840; summary(::IOContext{REPL.Terminals.TTYTerminal}, ::Array{Float64,1}, ::Tuple{Base.OneTo{Int64}})
                      1 .\show.jl:1893; showarg
                       1 .\strings\io.jl:31; print(::IOContext{REPL.Terminals.TTYTerminal}, ::Type)
                        1 .\show.jl:436; show(::IOContext{REPL.Terminals.TTYTerminal}, ::DataType)
                         1 .\show.jl:526; show_datatype(::IOContext{REPL.Terminals.TTYTerminal}, ::DataType)
                          1 .\show.jl:436; show(::IOContext{REPL.Terminals.TTYTerminal}, ::DataType)
                           1 .\show.jl:532; show_datatype(::IOContext{REPL.Terminals.TTYTerminal}, ::DataType)
                            1 .\show.jl:458; show_type_name(::IOContext{REPL.Terminals.TTYTerminal}, ::Core.TypeName)
                             1 .\strings\io.jl:156; string
                              1 .\strings\io.jl:121; print_to_string(::Symbol)
                               1 .\none:0; Type
                                1 .\iobuffer.jl:114; #IOBuffer#316(::Bool, ::Bool, ::Nothing, ::Bool, ::Int64, ::Int64, ::Type)
                                 1 .\strings\string.jl:71; StringVector
1421 .\task.jl:259; (::getfield(REPL, Symbol("##26#27")){REPL.REPLBackend})()
 1421 ...dbot\worker\package_win64\build\usr\share\julia\stdlib\v1.1\REPL\src\REPL.jl:117; macro expansion
  1421 ...dbot\worker\package_win64\build\usr\share\julia\stdlib\v1.1\REPL\src\REPL.jl:85; eval_user_input(::Any, ::REPL.REPLBackend)
   1421 .\boot.jl:328; eval(::Module, ::Any)
    1420 C:\Users\aporter\.julia\packages\NIDAQ\BztQ2\src\analog.jl:71; read(::NIDAQ.AITask, ::DataType, ::Int64)
     1420 C:\Users\aporter\.julia\packages\NIDAQ\BztQ2\src\properties.jl:185; getproperties
      1420 C:\Users\aporter\.julia\packages\NIDAQ\BztQ2\src\properties.jl:185; #getproperties#35
       3    C:\Users\aporter\.julia\packages\NIDAQ\BztQ2\src\properties.jl:101; _getproperties(::Tuple{Ptr{Nothing}}, ::String, ::Bool)
        3 .\none:0; #names
         3 .\reflection.jl:98; #names#7
          3 .\sort.jl:682; sort!
           3 .\sort.jl:694; #sort!#7
            3 .\sort.jl:634; sort!
             3 .\sort.jl:539; sort!
              2 .\sort.jl:545; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{Sym...
               1 .\sort.jl:545; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{Sy...
                1 .\sort.jl:545; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{Sy...
                 1 .\sort.jl:546; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{S...
                  1 .\sort.jl:546; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{S...
                   1 .\sort.jl:546; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{...
                    1 .\sort.jl:545; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array...
                     1 .\sort.jl:545; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array...
                      1 .\sort.jl:557; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Arra...
                       1 .\ordering.jl:49; lt
                        1 .\strings\basic.jl:319; isless
                         1 .\strings\basic.jl:317; cmp
                          1 .\c.jl:234; unsafe_convert
                           1 .\pointer.jl:58; unsafe_convert
               1 .\sort.jl:546; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{Sy...
                1 .\sort.jl:546; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{Sy...
                 1 .\sort.jl:546; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{S...
                  1 .\sort.jl:545; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{S...
                   1 .\sort.jl:546; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{...
                    1 .\sort.jl:545; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array...
                     1 .\sort.jl:545; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array...
                      1 .\sort.jl:557; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Arra...
                       1 .\ordering.jl:49; lt
                        1 .\strings\basic.jl:319; isless
                         1 .\strings\basic.jl:317; cmp
              1 .\sort.jl:546; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{Sym...
               1 .\sort.jl:546; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{Sy...
                1 .\sort.jl:545; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{Sy...
                 1 .\sort.jl:546; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{S...
                  1 .\sort.jl:545; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{S...
                   1 .\sort.jl:545; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array{...
                    1 .\sort.jl:546; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array...
                     1 .\sort.jl:545; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Array...
                      1 .\sort.jl:546; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Arra...
                       1 .\sort.jl:546; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Arra...
                        1 .\sort.jl:540; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.MergeSortAlg, ::Base.Order.ForwardOrdering, ::Arr...
                         1 .\sort.jl:455; sort!(::Array{Symbol,1}, ::Int64, ::Int64, ::Base.Sort.InsertionSortAlg, ::Base.Order.ForwardOrdering)
                          1 .\ordering.jl:49; lt
                           1 .\strings\basic.jl:319; isless
                            1 .\strings\basic.jl:317; cmp
       1409 C:\Users\aporter\.julia\packages\NIDAQ\BztQ2\src\properties.jl:102; _getproperties(::Tuple{Ptr{Nothing}}, ::String, ::Bool)
        1403 C:\Users\aporter\.julia\packages\NIDAQ\BztQ2\src\NIDAQ.jl:21; eval
         1403 .\boot.jl:328; eval
       5    C:\Users\aporter\.julia\packages\NIDAQ\BztQ2\src\properties.jl:103; _getproperties(::Tuple{Ptr{Nothing}}, ::String, ::Bool)
        2 .\strings\basic.jl:229; *
         1 .\strings\substring.jl:180; string(::String, ::String)
          1 .\strings\string.jl:60; _string_n
           1 .\essentials.jl:355; cconvert
            1 .\number.jl:7; convert
             1 .\boot.jl:738; Type
              1 .\boot.jl:708; toUInt64
               1 .\boot.jl:597; check_top_bit
         1 C:\Users\aporter\AppData\Local\Julia-1.1.0\lib\julia\sys.dll:?; #s5#3(::Any, ::Any, ::Any, ::Any)
        1 .\strings\io.jl:156; string
         1 .\strings\io.jl:123; print_to_string(::Symbol)
          1 .\show.jl:158; print
           1 .\io.jl:509; write
            1 .\iobuffer.jl:414; unsafe_write(::Base.GenericIOBuffer{Array{UInt8,1}}, ::Ptr{UInt8}, ::UInt64)
        1 .\strings\string.jl:247; getindex
       1    C:\Users\aporter\.julia\packages\NIDAQ\BztQ2\src\properties.jl:109; _getproperties(::Tuple{Ptr{Nothing}}, ::String, ::Bool)
        1 .\refpointer.jl:84; Ref{Bool32}(::Int64)
         1 .\refvalue.jl:8; Base.RefValue{Bool32}(::Int64)
          1 .\number.jl:7; convert(::Type{Bool32}, ::Int64)
       2    C:\Users\aporter\.julia\packages\NIDAQ\BztQ2\src\properties.jl:155; _getproperties(::Tuple{Ptr{Nothing}}, ::String, ::Bool)
        2 .\array.jl:705; iterate
         1 .\array.jl:729; getindex
         1 .\array.jl:199; length
```

### LINKED ISSUE #17: cannot add aichannels past ai7
When trying to add channels past ai7, i get this error:
```
julia> t = analog_input("Dev1/ai10")
ERROR: NIDAQmx: Requested value is not a supported value for this property. The property value may be invalid because it conflicts with another property.
```
I believe this is caused by:
```
Ref(codeunits(channel),1)
```
I have tested changing this to simply pointer(channel) and pointer(""). Will do a pull request once I'm finished breaking, then fixing.

### PR DESCRIPTION
**Pull request by alexporter8013**
Did some quick fixes for the digital and analog instantiation and reading. Also changed the terminal config dict to an enum to make the code a bit cleaner.
Fixes #16 
Fixes #17

### CODE REVIEW

File: src/NIDAQ.jl [hunk 1]
```diff
@@ -42,6 +42,8 @@ export analog_input_channels, analog_output_channels
 export digital_input_channels, digital_output_channels
 export counter_input_channels, counter_output_channels
 
+export TerminalConfig, RSE, NRSE, Differential, PseudoDifferential
```
Comment (bjarthur) [inline src/NIDAQ.jl]: will `TerminalConfig` ever need to be referenced by the user?  if not, perhaps don't export it.

File: src/analog.jl [hunk 2]
```diff
@@ -11,22 +8,22 @@ analog_input_configs = Dict{AbstractString,Number}(
 
 create an analog input channel, and a new task if one is not specified
 """
-function analog_input(channel::String; terminal_config="differential", range=nothing)
+function analog_input(channel::String; terminal_config::TerminalConfig=NRSE, range=nothing)
```
Comment (bjarthur) [inline src/analog.jl]: changing the default terminal configuration would be breaking, so we'd have to bump the minor version number.  is there a reason you advocate for this change?  what is the default when using LabView?

File: src/analog.jl [hunk 3]
```diff
@@ -11,22 +8,22 @@ analog_input_configs = Dict{AbstractString,Number}(
...
         device::String = split(channel,'/')[1]
         range=float(analog_input_ranges(device)[end,:])
     end
     catch_error( CreateAIVoltageChan(t.th,
-            Ref(codeunits(channel),1),
-            Ref(codeunits(""), 1),
-            analog_input_configs[terminal_config],
+            pointer(channel),
+            pointer(""),
+            Cuint(terminal_config),
```
Comment (bjarthur) [inline src/analog.jl]: could the the `Cuint()` caste be omitted here if `TerminalConfig` was declared to be `::Cuint` when defined?

File: src/analog.jl [hunk 4]
```diff
@@ -67,8 +64,10 @@ read_analog_cfunctions = Dict{Type,Function}(
     UInt16 => ReadBinaryU16,
     UInt32 => ReadBinaryU32 )
 
-function read(t::AITask, precision::DataType, num_samples_per_chan::Integer = -1)
-    num_channels = getproperties(t)["NumChans"][1]
+function read(t::AITask, num_samples_per_chan::Integer = -1, precision::DataType = Float64)
```
Comment (bjarthur) [inline src/analog.jl:67]: why the change in the order of the input arguments?  it's another breaking change.

File: src/digital.jl [hunk 5]
```diff
@@ -8,8 +8,8 @@ for (cfunction, jfunction, ntask) in (
     end
     @eval function $jfunction(t::$ntask, channel::String)
         catch_error( $cfunction(t.th,
-                Ref(codeunits(channel),1),
-                Ref(codeunits(""),1),
+                pointer(channel),
+                pointer(""),
```
Comment (bjarthur) [inline src/digital.jl]: in #17 you hypothesized that the the error when trying to add `ai10` was due to the `Ref` here, and that changing to `pointer` fixed it.  but is that really the case, or was using `NRSE` instead of `Differential` what was needed?  i ask, because `Ref` is "generally preferred" to `pointer` according to the latter's docstring.

**Review by bjarthur — COMMENTED** [review]
thanks for this!  definitely a good catch on the slowness of `getproperties`.  and i love the use of `@enum`; should've thought of that myself.

i've interspersed a few questions on specific lines.  looking forward to hearing back from you.

also note that if in your original comment when submitting this PR you include the phrase "fixes #16 and #17 " then those issues will be automatically closed when this PR is merged.

File: src/NIDAQ.jl (hunk 1 above)
Comment (alexporter8013) [reply src/NIDAQ.jl]: Will roll this in tomorrow when I get to work

File: src/analog.jl (hunk 2 above)
Comment (alexporter8013) [reply src/analog.jl]: I guess I'm not sure. Differential as default is probably less troublesome. Will roll that in. As an aside though, isn't changing the signature already breaking?

File: src/analog.jl (hunk 3 above)
Comment (alexporter8013) [reply src/analog.jl]: Will try that

File: src/analog.jl (hunk 4 above)
Comment (alexporter8013) [reply src/analog.jl:67]: Change in order was due to the reasoning that someone would be more likely to have the default datatype be Float64 than for someone to have the default n_samples. Therefore you get the convenience of specifying datatype only when it's different instead of any time the datatype and/or n_samples is different. I was actually considering just dropping support for the other data types in the higher level call and just leaving the lower level calls if someone wanted to be weird :stuck_out_tongue_winking_eye:

File: src/digital.jl (hunk 5 above)
Comment (alexporter8013) [reply src/digital.jl]: I was using the 144 analog input card. So it was kind of weird to be stopped after so few. I think the reason it was breaking was between the code units and ref. I think keeping these as pointers is safe in the context of the function call since they aren't being modified, or even declared in the function body. They shouldn't be garbage collected in the middle of a function call since they're arguments. But that's me speaking out of ignorance to how exactly the GC works.

I vote for keeping these specifically as pointers due to the aforementioned reasons.

Comment (bjarthur) [reply src/digital.jl]: @Cody-G experienced intermittent [problems](https://github.com/JaneliaSciComp/NIDAQ.jl/pull/15#issuecomment-457632258) with `pointer` which were fixed by changing to `Ref`.

would be nice to know for sure what your problem was before changing this.

Comment (alexporter8013) [reply src/digital.jl]: Interesting results. It seems only a certain range of values fail:
```julia> t0 = analog_input("Dev1/ai0", terminal_config=Differential)
NIDAQ.AITask(Ptr{Nothing} @0x000000001f20f900)

julia> t1 = analog_input("Dev1/ai10", terminal_config=Differential)
ERROR: NIDAQmx: Requested value is not a supported value for this property. The property value may be invalid because it conflicts with another property.
Stacktrace:
 [1] error(::String) at .\error.jl:33
 [2] #catch_error#1(::typeof(error), ::Function, ::Int32, ::String) at C:\Users\aporter\.julia\dev\NIDAQ\src\NIDAQ.jl:104
 [3] catch_error at C:\Users\aporter\.julia\dev\NIDAQ\src\NIDAQ.jl:97 [inlined] (repeats 2 times)
 [4] #analog_input#5(::TerminalConfig, ::Nothing, ::Function, ::NIDAQ.AITask, ::String) at C:\Users\aporter\.julia\dev\NIDAQ\src\analog.jl:23
 [5] #analog_input#4 at .\none:0 [inlined]
 [6] (::getfield(NIDAQ, Symbol("#kw##analog_input")))(::NamedTuple{(:terminal_config,),Tuple{TerminalConfig}}, ::typeof(analog_input), ::String) at .\none:0
 [7] top-level scope at none:0

julia> t2 = analog_input("Dev1/ai20", terminal_config=Differential)
NIDAQ.AITask(Ptr{Nothing} @0x000000001f20fc70)
```
Here is the result of me going from 0 - 143. Sorry for the bloat. I can't seem to upload files to comments.
```
julia> for i=0:143
           try
               t = analog_input("Dev1/ai$i", terminal_config=Differential)
               println("$i passed")
           catch
               println("$i failed")
           end
       end
0 passed
1 passed
2 passed
3 passed
4 passed
5 passed
6 passed
7 passed
8 failed
9 failed
10 failed
11 failed
12 failed
13 failed
14 failed
15 failed
16 passed
17 passed
18 passed
19 passed
20 passed
21 passed
22 passed
23 passed
24 failed
25 failed
26 failed
27 failed
28 failed
29 failed
30 failed
31 failed
32 passed
33 passed
34 passed
35 passed
36 passed
37 passed
38 passed
39 passed
40 failed
41 failed
42 failed
43 failed
44 failed
45 failed
46 failed
47 failed
48 passed
49 passed
50 passed
51 passed
52 passed
53 passed
54 passed
55 passed
56 failed
57 failed
58 failed
59 failed
60 failed
61 failed
62 failed
63 failed
64 passed
65 passed
66 passed
67 passed
68 passed
69 passed
70 passed
71 passed
72 failed
73 failed
74 failed
75 failed
76 failed
77 failed
78 failed
79 failed
80 passed
81 passed
82 passed
83 passed
84 passed
85 passed
86 passed
87 passed
88 failed
89 failed
90 failed
91 failed
92 failed
93 failed
94 failed
95 failed
96 passed
97 passed
98 passed
99 passed
100 passed
101 passed
102 passed
103 passed
104 failed
105 failed
106 failed
107 failed
108 failed
109 failed
110 failed
111 failed
112 passed
113 passed
114 passed
115 passed
116 passed
117 passed
118 passed
119 passed
120 failed
121 failed
122 failed
123 failed
124 failed
125 failed
126 failed
127 failed
128 passed
129 passed
130 passed
131 passed
132 passed
133 passed
134 passed
135 passed
136 failed
137 failed
138 failed
139 failed
140 failed
141 failed
142 failed
143 failed
```

Is there something I am missing?

Comment (alexporter8013) [reply src/digital.jl]: Here we go:
image
I am attempting to use invalid configurations. I will roll back to using Ref.

File: src/analog.jl (hunk 3 above)
Comment (alexporter8013) [reply src/analog.jl]: That worked.

File: src/analog.jl (hunk 4 above)
Comment (bjarthur) [reply src/analog.jl:67]: agree that changing the order makes sense, and having a default for precision is great too

File: src/analog.jl [hunk 6]
```diff
@@ -52,8 +49,8 @@ function analog_output(t::AOTask, channel::String; range=nothing)
         range=float(analog_output_ranges(device)[end,:])
     end
     catch_error( CreateAOVoltageChan(t.th,
-            Ref(codeunits(channel),1),
-            Ref(codeunits(""),1),
+            pointer(channel),
+            pointer(""),
```
Comment (bjarthur) [inline src/analog.jl]: change this back to Ref too?

File: src/digital.jl [hunk 7]
```diff
@@ -8,8 +8,8 @@ for (cfunction, jfunction, ntask) in (
     end
     @eval function $jfunction(t::$ntask, channel::String)
         catch_error( $cfunction(t.th,
-                Ref(codeunits(channel),1),
-                Ref(codeunits(""),1),
+                Ref(codeunits(channel)),
+                Ref(codeunits("")),
```
Comment (bjarthur) [inline src/digital.jl]: the second arg to Ref might default to 1 if the first is an array.  but i'm not sure.  should probably put it back in.

**Review by bjarthur — COMMENTED** [review]
looks great!  thanks for investigating the 144 channels issue.  just a couple small comments.

File: src/analog.jl (hunk 6 above)
Comment (alexporter8013) [reply src/analog.jl]: I thought I did... On my fork it shows that its back to Ref. Do I have to do anything special to re-sync?

File: src/digital.jl (hunk 7 above)
Comment (alexporter8013) [reply src/digital.jl]: I fixed this, in testing, the 1 is necessary.

File: src/analog.jl (hunk 6 above)
Comment (bjarthur) [reply src/analog.jl]: `git push --force ...` should do it.  or perhaps you forgot this one.  there are three places total that you originally made the change.

Comment (bjarthur) [reply src/analog.jl]: i just checked your [fork](https://github.com/bomber8013/NIDAQ.jl/blob/master/src/analog.jl#L53).  it still has `pointer`.

also, could you please squash your last four commits so as not to clutter the git log?  nominally you have made four changes here, which ideally would correspond to four commits:  1. fix the slow get_props, 2. use an enum, 3. change the order of the input args, and 4. add a default precision.  combining the last three into one is reasonable though.

Comment (bjarthur) [reply src/analog.jl]: and thanks for this by the way!  it's great to have some help.

Comment (alexporter8013) [reply src/analog.jl]: OK, I missed that last one apparently. I don't know how to squash commits currently. 

I am happy to help! I was just told by my manager that I shouldn't be letting my company's efforts just go for free into the world so you may no get anymore of these from me, despite the development I plan on doing, unfortunately.

**Comment by bjarthur** [comment]
i cleaned up the git commit history and merged your changes [manually](https://github.com/JaneliaSciComp/NIDAQ.jl/commit/190fb1d3b692e2663add93f57eae12431eae3d74).  was able to retain your name on two of the three commits.  thanks for the help!
